### PR TITLE
Improvements to sort functions

### LINKED ISSUE #291: Sort in OpenCL is too slow
I followed the suggestion on the page of 'sort segfault #233':
1. Updated the repositories
2. Used the develop branch of the 'compute' project

The previous error disappeared but it was significantly slower than before:
1.
array sortedDistances;
array indices;
sort(sortedDistances, indices, distances, 0);
took 204 seconds

where as
sortedDistances = sort(distances, 0);
took 4 seconds

2. Had two failed tests while the develop branch of 'compute' was used:
105 - Test_sort_by_key_opencl (Failed)
111 - Test_sort_index_opencl(Failed)

### PR DESCRIPTION
**Pull request by shehzan10**
This pull request adds:
* PERF: Improve the performance of batched sort along 0th dimension
* FEAT: Add sorting along higher dimensions
* Enable larger tests for sorting functions
* FEAT: Add multi dimension support for media
* Add multi dimensions tests for median

Fixes #291 #394

### CODE REVIEW

**Comment by shehzan10** [comment]
build arrayfire windows ci

**Comment by shehzan10** [comment]
@pavanky updated sort_index code for CPU and CUDA

**Comment by shehzan10** [comment]
build arrayfire tegrak1 ci
build arrayfire tegrax1 ci
build arrayfire linux-03 ci

**Comment by shehzan10** [comment]
build arrayfire tegrak1 ci

**Comment by shehzan10** [comment]
build arrayfire linux-03 ci

File: src/backend/cuda/kernel/sort_by_key.hpp [hunk 1]
```diff
@@ -44,14 +44,98 @@ namespace cuda
...
+            pSeq.strides[0] = 1;
+            pSeq.dims[0] = keydims[0];
+            for(int i = 1; i < 4; i++) {
+                pSeq.dims[i] = keydims[i];
+                pSeq.strides[i] = pSeq.strides[i - 1] * pSeq.dims[i - 1];
+            }
+            cuda::kernel::iota<uint>(pSeq, seqDims, tileDims);
+
+            // Make pkey, pVal into a pair
+            thrust::device_vector<IndexPair<Tk, Tv> > X(inDims.elements());
```
Comment (pavanky) [inline src/backend/cuda/kernel/sort_by_key.hpp]: Allocate using arrayfire. This is going to be costly otherwise.

**Comment by shehzan10** [comment]
@pavanky @9prady9 Can you guys review https://github.com/arrayfire/arrayfire/pull/1373/commits/f6eae071675f8fe4f2c5672b104a3c1394de66b7

**Comment by 9prady9** [comment]
@shehzan10 Tests look fine.

**Comment by pavanky** [comment]
looks good :+1:

**Comment by pavanky** [comment]
build arrayfire ci

**Comment by shehzan10** [comment]
build arrayfire ci

**Pull request merged**